servers/rendering/rendering_device_commons.h

```diff
@@ -271,6 +271,28 @@ class RenderingDeviceCommons : public Object {
 		DATA_FORMAT_MAX,
 	};
 
+	enum BreadcrumbMarker {
+		NONE = 0,
+		// Environment
+		REFLECTION_PROBES,
+		SKY_PASS,
+		// Light mapping
+		LIGHTMAPPER_PASS,
+		// Shadows
+		SHADOW_PASS_DIRECTIONAL,
+		SHADOW_PASS_CUBE,
+		// Geometry passes
+		OPAQUE_PASS,
+		ALPHA_PASS,
+		TRANSPARENT_PASS,
+		// Screen effects
+		POST_PROCESSING_PASS,
+		BLIT_PASS,
+		UI_PASS,
+		// Other
+		DEBUG_PASS
+	};
+
```

Comment:
I feel the responsibility of RenderingDevice should stay as abstract as possible in this case and not have knowledge of how the rendering pipeline works. Perhaps we can keep the breadcrumbs as a typedef of just `uint32_t` and the renderers (Forward+, Mobile, or future ones) can assign the semantic meaning to the numbers?

Reply:
Breadcrumbs don't really add knowledge.

They're just tagging an ID. The idea is that if the app crashes we look at the report and see "what was being executed?" oh breadcrumb says ID = 2, which corresponds to... let me see... SKY_PASS.

It's up to the rendering pipelines to _actually_ use those IDs.

In fact some parts of the code do the following:

```
breadcrumb = LIGHTMAPPER_PASS | (pass_id << 16)
```

In other words, it doesn't give knowledge about the upper levels; it's just a way of standardizing some common tags that are more likely to be used.

Reply:
But wouldn't that mean that:

1. Mobile and Forward+ would have their own copies of the enum, which can get out of sync. The main issue about that is that it can cause silly mistakes when debugging that can waste hours (i.e. user got id = 7, dev looked it up what that means, but saw the wrong version).
2. Users who want to use the `RenderingDevice` directly have no information about what that id is for, and even if they do, they would not adhere to any convention.

Btw I looked again and turns out only 16 bits of each breadcrumb argument are used:

```cpp
uint32_t breadcrumb = (p_phase << 16) | (p_breadcrumb_data & ((1 << 16) - 1));
```

Perhaps we could change this argument signature to...?

// original:
void draw_list_begin(RID p_framebuffer, InitialAction p_initial_color_action, FinalAction p_final_color_action, InitialAction p_initial_depth_action, FinalAction p_final_depth_action, const Vector<Color> &p_clear_color_values, float p_clear_depth, uint32_t p_clear_stencil, const Rect2 &p_region, RDD::BreadcrumbMarker p_phase, uint32_t p_breadcrumb_data) {}

// My proposal:
enum BreadcrumbMarker {
   // Stays the same.
}

uint32_t breadcrumb( BreadcrumbMarker p_phase, uint16_t p_data ) {
  return (p_phase << 16) | (p_data & 0xFFFF)
}

void draw_list_begin(RID p_framebuffer, InitialAction p_initial_color_action, FinalAction p_final_color_action, InitialAction p_initial_depth_action, FinalAction p_final_depth_action, const Vector<Color> &p_clear_color_values, float p_clear_depth, uint32_t p_clear_stencil, const Rect2 &p_region, uint32_t p_breadcrumb_data) {}
```

This way the BreadcrumbMarker enum is still visible to everyone, but it is made clear the underlying datatype is an arbitrary uint32_t.

Reply:
Looking back on this. I like your suggested approach. I really don't love the RDD having high level concepts like `POST_PROCESSING_PASS` or `SHADOW_PASS_CUBE`, but at the same time, to make this feature useful we want the debug utility to print out the human readable name for the pass and not force users to look it up. You are right that looking up an ID would be a huge annoyance and significantly reduce the value of breadcrumbs.